### PR TITLE
feat(S3_in_w_x_flags): Support S3 URIs for custom checks paths and whitelist files.

### PR DESCRIPTION
**Pull request by MrCloudSec**
### Context 

This new feature will allow users to input S3 URIs when working with whitelists (-w flag) or custom checks paths (-x flag).

### Description

If an user input a S3 URI in flags -w or -x, Prowler will automatically read the files from the indicated S3 bucket.
For example:
./prowler -c extra764 -p <profile_name> -w s3://bucket/whitelist.txt
./prowler -c extra764 -p <profile_name> -x s3://bucket/checks

### License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -292,11 +292,12 @@ Prowler has two parameters related to regions: `-r` that is used query AWS servi
 
     >Note about output formats to use with `-M`: "text" is the default one with colors, "mono" is like default one but monochrome, "csv" is comma separated values, "json" plain basic json (without comma between lines) and "json-asff" is also json with Amazon Security Finding Format that you can ship to Security Hub using `-S`.
 
-    or save your report in an S3 bucket (this only works for text or mono. For csv, json or json-asff it has to be copied afterwards):
+    To save your report in an S3 bucket, use `-B` to define a custom output bucket along with `-M` to defiine the output format that is going to be updated to S3:
```
Comment (toniblyx) [inline README.md]:
```suggestion
    To save your report in an S3 bucket, use `-B` to define a custom output bucket along with `-M` to define the output format that is going to be uploaded to S3:
```

File: README.md [hunk 2]
```diff
@@ -399,7 +400,9 @@ Prowler runs in GovCloud regions as well. To make sure it points to the right AP
 
 ### Custom folder for custom checks
 
-Flag `-x /my/own/checks` will include any check in that particular directory. To see how to write checks see [Add Custom Checks](#add-custom-checks) section.
+Flag `-x /my/own/checks` will include any check in that particular directory (files must start by check). To see how to write checks see [Add Custom Checks](#add-custom-checks) section.
+
+S3 URIs are also supported as custom folders for custom checks, e.g. `s3://bucket/prefix/checks`
```
Comment (toniblyx) [inline README.md]:
```suggestion
S3 URIs are also supported as custom folders for custom checks, e.g. `s3://bucket/prefix/checks`. Prowler will download the folder locally and run the checks as they are called with default execution,`-c` or `-g`.
```

File: include/custom_checks [hunk 3]
```diff
@@ -0,0 +1,52 @@
+#!/usr/bin/env bash
```
Comment (toniblyx) [inline include/custom_checks:1]: every message that starts by $OK or $BAD to colorize the output must be clean up after each sentence with $NORMAL. So, for instance `echo -e "$OK Success! Custom checks were downloaded and included, starting Prowler..."` must actually be `echo -e "$OK Success! Custom checks were downloaded and included, starting Prowler...$NORMAL"`

**Review by toniblyx — CHANGES_REQUESTED** [review]

File: include/allowlist [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+# under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+# CONDITIONS OF ANY KIND, either express or implied. See the License for the
+# specific language governing permissions and limitations under the License.
+
+allowlist(){
+  # check if the file is an S3 URI
+  if grep -q -E "^s3://([^/]+)/(.*?([^/]+))$" <<< "$ALLOWLIST_FILE"; then
+    # download s3 object
+    echo -e "$NOTICE Downloading allowlist from S3 URI "$ALLOWLIST_FILE" ..."  
+    if ! $AWSCLI s3 cp $ALLOWLIST_FILE allowlist_s3_file.txt $PROFILE_OPT > /dev/null 2>&1; then
```
Comment (jfagoagas) [inline include/allowlist]: Please, include `allowlist_s3_file.txt` as a local variable like:
```sh
local ALLOWLIST_S3_FILE=allowlist_s3_file.txt
```

Maybe you can use `ALLOWLIST_S3_FILE` and `ALLOWLIST_LOCAL_FILE`

File: include/allowlist [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+      echo "$BAD FAIL! Access Denied trying to download allowlist from the S3 URI, please make sure it is correct and/or you have permissions to get the S3 object.$NORMAL"
+      EXITCODE=1
+      exit $EXITCODE
+    fi
+    echo -e "$OK Success! Allowlist was downloaded, starting Prowler...$NORMAL"
+    ALLOWLIST_FILE=allowlist_s3_file.txt
```
Comment (jfagoagas) [inline include/allowlist]: This temporary file could be a local variable

File: include/allowlist [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    # ignore inline comments: check1:foo # inline comment
+    ALLOWLIST=$(awk '!/^[[:space:]]*#/{print }' <(cat "$ALLOWLIST_FILE") | sed 's/[[:space:]]*#.*$//g')
+    # remove temporary file
+    rm -f "$ALLOWLIST_FILE"
+  else
+    # Check if input allowlist file exists
+    if [[ -f "$ALLOWLIST_FILE" ]]; then
+      # ignore lines starting with # (comments)
+      # ignore inline comments: check1:foo # inline comment
+      ALLOWLIST=$(awk '!/^[[:space:]]*#/{print }' <(cat "$ALLOWLIST_FILE") | sed 's/[[:space:]]*#.*$//g')
```
Comment (jfagoagas) [inline include/allowlist:36]: Think about the following logic:
1. Check if there is an input whitelist, from a local file or from S3
2. Load `ALLOWLIST` variable with the allowed resources.

Only an opinion to reduce some lines.

File: include/custom_checks [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software distributed
+# under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+# CONDITIONS OF ANY KIND, either express or implied. See the License for the
+# specific language governing permissions and limitations under the License.
+
+custom_checks(){
+  # check if the path is an S3 URI
+  if grep -q -E "^s3://([^/]+)/?(.*?([^/]+)/?)?$" <<< "$EXTERNAL_CHECKS_PATH"; then
+    if grep -q "check*" <<< "$($AWSCLI s3 ls "$EXTERNAL_CHECKS_PATH" $PROFILE_OPT)"; then
```
Comment (jfagoagas) [inline include/custom_checks]:
```suggestion
    if grep -q "check*" <<< "$("${AWSCLI}" s3 ls "${EXTERNAL_CHECKS_PATH}" ${PROFILE_OPT})"; then
```

File: include/custom_checks [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+      mkdir "${S3_CHECKS_TEMP_FOLDER}"
+      $AWSCLI s3 sync "$EXTERNAL_CHECKS_PATH" "${S3_CHECKS_TEMP_FOLDER}" $PROFILE_OPT > /dev/null
+      # verify if there are checks
+      for checks in "${S3_CHECKS_TEMP_FOLDER}"/check*; do
+        . "$checks"
+        echo -e "$OK Check $(basename "$checks") was included!$NORMAL"
+      done
+    echo -e "$OK Success! Custom checks were downloaded and included, starting Prowler...$NORMAL"
+    # remove temporary dir
+    rm -rf "${S3_CHECKS_TEMP_FOLDER}"
```
Comment (jfagoagas) [inline include/custom_checks:30]: Can you remove the custom checks folder prior to start the execution of this checks?

File: include/custom_checks [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+      # remove temporary dir
+      rm -rf "${S3_CHECKS_TEMP_FOLDER}"
+      exit $EXITCODE
+    fi
+  else
+    # verify if input directory exists with checks
+    if ls "${EXTERNAL_CHECKS_PATH}"/check* > /dev/null 2>&1; then
+      for checks in "${EXTERNAL_CHECKS_PATH}"/check*; do
+        . "$checks"
+        echo -e "$OK Check $(basename "$checks") was included!$NORMAL"
```
Comment (jfagoagas) [inline include/custom_checks:43]: Same here, think about first recovering checks and then a unique load.

**Review by jfagoagas — CHANGES_REQUESTED** [review]
Good job! Please check my comments above.

As a general comment, review variable braces `{}`

**Comment by jfagoagas** [comment]
Do we need to include required AWS S3 API permissions?

**Comment by MrCloudSec** [comment]
> Do we need to include required AWS S3 API permissions?

Review the last commit with those changes pls.

**Review by jfagoagas — APPROVED** [review]
🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 

New features!! 👏

**Review by toniblyx — APPROVED** [review]
LGTM

**Pull request merged**